provider/pkg/provider/remote/copyController.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"errors"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/sftp"
+
+	p "github.com/pulumi/pulumi-go-provider"
+	"github.com/pulumi/pulumi-go-provider/infer"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/diag"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+)
+
+// These are not required. They indicate to Go that Command implements the following interfaces.
+// If the function signature doesn't match or isn't implemented, we get nice compile time errors in this file.
+var _ = (infer.CustomResource[CopyInputs, CopyOutputs])((*Copy)(nil))
+
+func doCopy(ctx p.Context, input CopyInputs) (CopyOutputs, error) {
+	sourcePath := input.sourcePath()
+
+	ctx.Logf(diag.Debug,
+		"Creating remote: %s:%s from local '%s'",
+		input.Connection.Host, input.RemotePath, sourcePath)
+
+	client, err := input.Connection.Dial(ctx)
+	if err != nil {
+		return CopyOutputs{input}, err
+	}
+	defer client.Close()
+
+	sftp, err := sftp.NewClient(client)
+	if err != nil {
+		return CopyOutputs{input}, err
+	}
+	defer sftp.Close()
+
+	src, err := os.Open(sourcePath)
+	if err != nil {
+		return CopyOutputs{input}, err
+	}
+	defer src.Close()
+
+	srcInfo, err := src.Stat()
+	if err != nil {
+		return CopyOutputs{input}, err
+	}
+	if srcInfo.IsDir() {
+		err = copyDir(sftp, sourcePath, input.RemotePath)
+	} else {
+		err = copyFile(sftp, sourcePath, input.RemotePath)
+	}
+	return CopyOutputs{input}, err
+}
+
+// This is the Create method. This will be run on every Copy resource creation.
+func (*Copy) Create(ctx p.Context, name string, input CopyInputs, preview bool) (string, CopyOutputs, error) {
+	if preview {
+		return "", CopyOutputs{input}, nil
+	}
+
+	outputs, err := doCopy(ctx, input)
+	if err != nil {
+		return "", CopyOutputs{input}, err
+	}
+
+	id, err := resource.NewUniqueHex("", 8, 0)
+	return id, outputs, err
+}
+
+func (c *Copy) Update(ctx p.Context, id string, olds CopyOutputs, news CopyInputs, preview bool) (CopyOutputs, error) {
+	if err := news.validate(); err != nil {
+		return olds, err
+	}
+
+	needCopy := true
+	if news.hash() == olds.hash() && news.RemotePath == olds.RemotePath {
+		needCopy = false
+	}
+
+	if preview {
+		// TODO,tkappler how to show the user whether we need to copy or not?
```

Comment:
I was thinking of preview, since we're inside the `if preview` block. Shouldn't the preview show the user what will be done? I.e., will there be a fresh copy?

Reply:
I'm concerned it will be noisy, but I might be wrong. As I understand it, it's just adding or removing a couple of lines, so I would implement the logging and see if its too noisy. If you like the more verbose messaging better, then leave it in.

We can futz with the logging after GA, so I'll be happy with whatever you think best.